File: hena/tyama_hena24_enum.php
#!/usr/bin/env php
<?php
//http://qiita.com/Nabetani/items/1c83005a854d2c6cbb69
//http://nabetani.sakura.ne.jp/hena/ord24eliseq/

//function cbrt($n){
//	return pow($n,1-2.0/3);
//}

function isqrt($n){
	if($n<=0)return 0;
	if($n<4)return 1;
	$x=0;$y=$n;
	for(;$x!=$y&&$x+1!=$y;){$x=$y;$y=intval((intval($n/$y)+$y)/2);}
	return $x;
}
function icbrt($n){
	if($n<0)return icbrt(-$n);
	if($n==0)return 0;
	if($n<8)return 1;
	$x=0;$y=$n;
	for(;$x!=$y&&$x+1!=$y;){$x=$y;$y=intval( (intval(intval($n/$y)/$y)+$y*2)/3 );}
	return $x;
}

function generate(){
	$i=1;
	for(;;){
		yield $i;
		$i++;
	}
}

function drop_prev($check,$prev){
	$prev->rewind();
	$a=$prev->current();
	$prev->next();
	$b=$prev->current();
	for(;;){
		if(!$check($b))yield $a;
		$a=$b;
		$prev->next();
		$b=$prev->current();
	}
}
function drop_next($check,$prev){
	$prev->rewind();
	$a=$prev->current();
	$prev->next();
	$b=$prev->current();
	yield $a;
	for(;;){
		if(!$check($a))yield $b;
		$a=$b;
		$prev->next();
		$b=$prev->current();
	}
}
function drop_n($check,$n,$prev){
	$prev->rewind();
	$i=0;
	for(;;){
		$i+=1;
		$a=$prev->current();
		$prev->next();
		if(!$check($i,$n))yield $a;
	}
}
function is_sq($n){
	//$x=intval(sqrt($n));
	$x=isqrt($n);
	return $x*$x==$n;
}
function is_cb($n){
	//$x=intval(cbrt($n));
	$x=icbrt($n);
	return $x*$x*$x==$n;
}
function is_multiple($i,$n){
	return $i%$n==0;
}
function is_le($i,$n){
	return $i<=$n;
}

$f=[
	'S'=>function($enum){return drop_next('is_sq',$enum);},
	's'=>function($enum){return drop_prev('is_sq',$enum);},
	'C'=>function($enum){return drop_next('is_cb',$enum);},
	'c'=>function($enum){return drop_prev('is_cb',$enum);},
	'h'=>function($enum){return drop_n('is_le',100,$enum);},
];
for($i=2;$i<=9;$i++){$f[(string)$i]=function($enum)use($i){return drop_n('is_multiple',$i,$enum);};}

for(;$line=fgets(STDIN);){
	//cS => $f['S']($f['c'](generate()))
	$z=array_reduce(str_split(rtrim($line)),function($s,$e)use($f){return $f[$e]($s);},generate());
	$z->rewind();
	$cnt=10;
	$a=[];
	for($i=0;$i<$cnt;$i++){
		$a[]=$z->current();
		if($i<$cnt-1)$z->next();
	}
	echo implode(',',$a).PHP_EOL;
	flush(); //STDOUT);
}

File: hena/tyama_hena24_enum.jl
#!/usr/bin/env julia

# http://qiita.com/Nabetani/items/1c83005a854d2c6cbb69
# http://nabetani.sakura.ne.jp/hena/ord24eliseq/

function isqrt(n::Int)
	if n<=0
		return 0
	end
	if n<4
		return 1
	end
	local x=0
	local y=n
	while x!=y&&x+1!=y
		x=y
		y=(n÷y+y)÷2
	end
	return x
end

function icbrt(n::Int)
	if n<0
		return icbrt(-n)
	end
	if n==0
		return 0
	end
	if n<8
		return 1
	end
	local x=0
	local y=n
	while x!=y&&x+1!=y
		x=y
		y=(n÷y÷y+y*2)÷3
	end
	return x
end

function is_sq(n::Int)
	local x=isqrt(n)
	return x*x==n
end
function is_cb(n::Int)
	local x=icbrt(n)
	return x*x*x==n
end
is_multiple(i::Int,n::Int) = i%n==0
is_le(i::Int,n::Int) = i<=n

function generate()
	local ch = Channel{Int}(8)
	function _producer()
		local i=1
		while true
			put!(ch,i)
			i+=1
		end
	end
	@async _producer()
	ch
end

function drop_prev(check::Function,prev::Channel{Int})
	local ch = Channel{Int}(8)
	function _producer()
		local a=take!(prev)
		local b=take!(prev)
		while true
			if !check(b)
				put!(ch,a)
			end
			a=b
			b=take!(prev)
		end
	end
	@async _producer()
	return ch
end

function drop_next(check,prev::Channel{Int})
	local ch = Channel{Int}(8)
	function _producer()
		local a=take!(prev)
		local b=take!(prev)
		put!(ch,a)
		while true
			if !check(a)
				put!(ch,b)
			end
			a=b
			b=take!(prev)
		end
	end
	@async _producer()
	return ch
end

function drop_n(check,n::Int,prev::Channel{Int})
	local ch = Channel{Int}(8)
	function _producer()
		local i=0
		while true
			i+=1
			a=take!(prev)
			if !check(i,n)
				put!(ch,a)
			end
		end
	end
	@async _producer()
	return ch
end

Base.@ccallable function julia_main(args::Vector{String})::Cint
	local f=Dict(
		'S' => prev -> drop_next(is_sq,prev),
		's' => prev -> drop_prev(is_sq,prev),
		'C' => prev -> drop_next(is_cb,prev),
		'c' => prev -> drop_prev(is_cb,prev),
		'h' => prev -> drop_n(is_le,100,prev)
	)
	for j in 2:9
		#ラムダ式のキャプチャはバグらないみたいですね
		f[j+'0']=prev -> drop_n(is_multiple,j,prev)
	end

	local line=readline(stdin)
	while line!=""
		line=rstrip(line)
		#cS => f['S'](f['c'](generate()))
		local ch = generate()
		for c in line
			ch=f[c](ch)
		end
		for i in 0:9
			if i>0
				print(",")
			end
			a=take!(ch)
			print(a)
		end
		print("\n")
		flush(stdout)
		line=readline(stdin)
	end
	return 0
end

if get(ENV, "COMPILE_STATIC", "false") == "false"
	julia_main(ARGS)
end
